### PR TITLE
feat: Add SpeculatorTrainer for draft model training

### PR DESCRIPTION
**Pull request by hrathina**
## Summary

* **New Features**
  * Added speculative-decoding (speculator) trainer support with new public exports and expanded RHAI trainer type coverage.
  * Supports configurable verifier/draft training, optional pip-install inputs, distributed execution, metrics endpoint reporting, and optional progression tracking.

* **Bug Fixes**
  * Added stricter validation for trainer configuration (mode/type, numeric ranges, allowed dtypes) and improved normalization/handling of `output_dir` and hidden-states URI schemes.
  * Updated volume mounting to deduplicate overlapping PVC mounts and error on mismatched PVCs.

* **Tests**
  * Added extensive unit tests for validation, training-script rendering, CR generation, progression instrumentation, and PVC mount behavior.

### CODE REVIEW

File: kubeflow/trainer/rhai/speculator.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        "    transform=AddGaussianNoise(),\n"
+        f"    hidden_states_dtype=torch.{trainer.hidden_states_dtype},\n"
+        ")\n"
+        "val_dataset = ArrowDataset(\n"
+        "    max_len=max_len,\n"
+        f"    datapath={trainer.hidden_states_path!r},\n"
+        "    split_ratio=-0.1,\n"
+        "    on_missing='skip',\n"
+        f"    hidden_states_dtype=torch.{trainer.hidden_states_dtype},\n"
+        ")\n"
```
[the remark this answers is not in the document]
Comment (hrathina) [reply kubeflow/trainer/rhai/speculator.py]: Resolved

**Review by efazal — COMMENTED** [review]
Good work Hari, solid start on the train_only mode. Few things to address:

File: kubeflow/trainer/rhai/speculator.py [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+        f"    save_path={save_path!r},\n"
+        "    is_distributed=is_distributed,\n"
+        "    local_rank=local_rank,\n"
+        "    train_call_kwargs={'shift_fn': shift_batch},\n"
+        "    val_call_kwargs={'shift_fn': shift_batch},\n"
+        ")\n"
+        "\n"
+        "trainer = Trainer(model, config, train_loader, val_loader)\n"
+        "trainer.run_training()\n"
+    )
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:841]: We shouldn't be returning whole functions as strings, it's fragile and not readable. We should use real python code that shows syntaxes. Follow the same pattern used in transformerTrainer code, where we use `inspect.getsource()`

File: kubeflow/trainer/rhai/speculator.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+
+    Args:
+        TRAIN_ONLY: Train draft model from pre-extracted hidden states on PVC.
+        DATA_ONLY: Extract hidden states from verifier model via vLLM (future).
+        OFFLINE: Train using a user-managed vLLM endpoint (future).
+    """
+
+    TRAIN_ONLY = "train_only"
+    DATA_ONLY = "data_only"
+    OFFLINE = "offline"
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:45]: Include `online` mode as well, user can do online training if they are using a custom runtime image with all dependencies

File: kubeflow/trainer/rhai/speculator.py [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+
+        if self.speculator_type not in _SUPPORTED_TYPES:
+            supported = ", ".join(t.value for t in _SUPPORTED_TYPES)
+            raise ValueError(
+                f"Speculator type '{self.speculator_type.value}' is not yet supported. "
+                f"Currently supported types: {supported}."
+            )
+
+        if self.mode == SpeculatorMode.TRAIN_ONLY and not self.hidden_states_path:
+            raise ValueError(
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:204]: should raise `NotImplementedError` instead of value error.

File: kubeflow/trainer/rhai/speculator.py [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+    EAGLE3 = "eagle3"
+    DFLASH = "dflash"
+    MTP = "mtp"
+    PEAGLE = "peagle"
+
+
+_SUPPORTED_MODES = {SpeculatorMode.TRAIN_ONLY}
+_SUPPORTED_TYPES = {SpeculatorType.EAGLE3}
+_SUPPORTED_DTYPES = {"bfloat16", "float16", "float32"}
+
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py:66]: Do we require the above supported modes and types? why even include them if they are not supported? it's misleading.

File: kubeflow/trainer/rhai/speculator.py [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+        "import torch\n"
+        "from torch.utils.data import DataLoader\n"
+        "from transformers import AutoConfig\n"
+        "from speculators.models.eagle3.core import Eagle3DraftModel\n"
+        "from speculators.train.trainer import Trainer, TrainerConfig\n"
+        "from speculators.train.data import (\n"
+        "    ArrowDataset,\n"
+        "    create_collate_fn,\n"
+        ")\n"
+        "from speculators.train.noise_transforms import AddGaussianNoise\n"
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: It should be `AddUniformNoise` instead of `AddGaussianNoise`

File: kubeflow/trainer/rhai/speculator.py [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+        "\n"
+        f"max_len = {trainer.total_seq_len}\n"
+        "collate_fn = create_collate_fn(max_len, verifier_config.hidden_size)\n"
+        "\n"
+        "train_dataset = ArrowDataset(\n"
+        "    max_len=max_len,\n"
+        f"    datapath={trainer.hidden_states_path!r},\n"
+        "    split_ratio=0.9,\n"
+        "    on_missing='skip',\n"
+        "    transform=AddGaussianNoise(),\n"
```
Comment (efazal) [inline kubeflow/trainer/rhai/speculator.py]: same comment as above

**Comment by hrathina** [comment]
@coderabbitai review

**Comment by hrathina** [comment]
@coderabbitai review

File: kubeflow/trainer/rhai/speculator.py [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+                    if (
+                        _steps_per_epoch is None
+                        and "train" in msg
+                        and "epoch" in msg
+                        and "global_step" in msg
+                    ):
+                        if msg["epoch"] == 0:
+                            _max_step_in_epoch0 = max(_max_step_in_epoch0, msg["global_step"])
+                        elif msg["epoch"] >= 1 and _max_step_in_epoch0 >= 0:
+                            _steps_per_epoch = _max_step_in_epoch0 + 1
```
[the remark this answers is not in the document]
Comment (hrathina) [reply kubeflow/trainer/rhai/speculator.py:930]: The _steps_per_epoch is already set correctly before training starts via _set_steps_per_epoch(len(train_loader)) (line 311), which handles all epoch counts including epochs=1. 

This was verified on cluster with epochs=1, progress tracked correctly from 0% to 100%. The suggested fix would actually break single-epoch training: on the first logged step, _max_step_in_epoch0 = 0, so _steps_per_epoch would be locked to 1, causing progress to jump to 100% after step 1 of 43.
